CMakeLists.txt

```diff
@@ -43,281 +29,65 @@ else()
 endif()
 
-if(USE_TBB)
-    find_package(TBB REQUIRED)
-    add_definitions(-DCNN_USE_TBB)
-    include_directories(${TBB_INCLUDE_DIRS})
-    link_directories(${TBB_LIBRARY_DIRS})
-    set(REQUIRED_LIBRARIES ${REQUIRED_LIBRARIES} ${TBB_LIBRARIES})
-elseif(USE_OMP)
-    find_package(OpenMP REQUIRED)
-    add_definitions(-DCNN_USE_OMP)
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
-endif()
-
-if(USE_NNPACK)
-   find_package(NNPACK REQUIRED)
-   add_definitions(-DCNN_USE_NNPACK)
-   include_directories(SYSTEM ${NNPACK_INCLUDE_DIR})
-   include_directories(SYSTEM ${NNPACK_INCLUDE_DIR}/../third-party/pthreadpool/include)
-   set(REQUIRED_LIBRARIES ${REQUIRED_LIBRARIES} ${NNPACK_LIB})
-endif()
+# Using cmake scripts and modules
+list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake/Modules)
 
-# ----------------------------------------------------------------------------
-# COMPILER OPTIONS
-# ----------------------------------------------------------------------------
+include(cmake/summary.cmake)
+include(cmake/config_gen.cmake)
 
-if(CMAKE_COMPILER_IS_GNUCXX OR MINGW OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-    if(USE_SSE)
-        add_definitions(-DCNN_USE_SSE)
-        set(EXTRA_C_FLAGS  " ${EXTRA_C_FLAGS} -msse3")
-    endif()
-    if(USE_AVX)
-        add_definitions(-DCNN_USE_AVX)
-        set(EXTRA_C_FLAGS  "${EXTRA_C_FLAGS} -mavx -mavx2 -march=core-avx2")
-    endif()
-    if((NOT USE_TBB) AND (NOT USE_OMP))
-        set(EXTRA_C_FLAGS " ${EXTRA_C_FLAGS} -pthread")
-    endif()
-    set(EXTRA_C_FLAGS " ${EXTRA_C_FLAGS} -std=c++11 -Wall -Wpedantic")
-    set(EXTRA_C_FLAGS_RELEASE " -O3")
-    set(EXTRA_C_FLAGS_DEBUG " -g3")
-else() # MSVC
-    if(USE_SSE)
-        add_definitions(-DCNN_USE_SSE)
-        set(EXTRA_C_FLAGS  " ${EXTRA_C_FLAGS} /arch:SSE2 ")
-    endif()
-    if(USE_AVX)
-        add_definitions(-DCNN_USE_AVX)
-        set(EXTRA_C_FLAGS  "${EXTRA_C_FLAGS}  /arch:AVX ")
-    endif()
-    set(EXTRA_C_FLAGS_RELEASE " /Ox /bigobj ")
-    add_definitions(-D _CRT_SECURE_NO_WARNINGS)
-endif()
+# ---[ Dependencies
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${EXTRA_C_FLAGS}")
-set(CMAKE_CXX_FLAGS_RELEASE "${EXTRA_C_FLAGS_RELEASE}")
-set(CMAKE_CXX_FLAGS_DEBUG "${EXTRA_C_FLAGS_DEBUG}")
-set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${EXTRA_EXE_LINKER_FLAGS}")
-set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE} ${EXTRA_EXE_LINKER_FLAGS_RELEASE}")
-set(CMAKE_EXE_LINKER_FLAGS_DEBUG "${CMAKE_EXE_LINKER_FLAGS_DEBUG} ${EXTRA_EXE_LINKER_FLAGS_DEBUG}")
+include(cmake/dependencies.cmake)
 
+# ---[ Compiler options
 
-#------------------------------------------------
-# TARGETS
-#------------------------------------------------
+include(cmake/flags.cmake)
 
+# ---[ Includes
 
-include_directories(${PROJECT_SOURCE_DIR})
-link_libraries(${REQUIRED_LIBRARIES})
+file(GLOB_RECURSE tiny_cnn_hdrs tiny_cnn/*.h)
```

Comment:
you have a point but globbing it's not that bad. Since the project is growing more and more it's not clean to have a huge manual list of files. IMO, there are solutions to work around this updating files issues: 1) Have as a good practice to run manually the cmake at developing time. 2) touching CMakeLists.txt